optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,350 @@
+                param_name, trial.number, self._independent_sampler.__class__.__name__
+            )
+        )
+
+    @staticmethod
+    def _log_asyncronous_seeding() -> None:
+        _logger.warning(
+            "No seed is provided for `QMCSampler` and the seed is set randomly. "
+            "If you are running multiple `QMCSampler`s in parallel and/or distributed "
+            " environment, the same seed must be used in all samplers to ensure that resulting "
+            "samples are taken from the same QMC sequence. "
+            "You can suppress this warning by setting `warn_asyncronous_seeding` "
+            "to `False` in the constructor of `QMCSampler`, "
+            "if this random seeding is intended behavior."
+        )
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+
+        if self._initial_search_space is not None:
+            if self._warn_independent_sampling:
+                self._log_independent_sampling(trial, param_name)
+
+        return self._independent_sampler.sample_independent(
+            study, trial, param_name, param_distribution
+        )
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> Dict[str, Any]:
+
+        if search_space == {}:
+            return {}
+
+        sample = self._sample_qmc(study, trial, search_space)
+        trans = _SearchSpaceTransform(search_space)
+        sample = scipy.stats.qmc.scale(sample, trans.bounds[:, 0], trans.bounds[:, 1])
+        sample = trans.untransform(sample[0, :])
+        return sample
+
+    def _sample_qmc(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> numpy.ndarray:
+
+        # Lazy import because the `scipy.stats.qmc` is slow to import.
+        import scipy.stats.qmc
+
+        sample_id = self._find_sample_id(study, trial, search_space)
+        d = len(search_space)
+
+        # Use cached `qmc_engine` or construct a new one.
+        if self._is_engine_cached(d, sample_id):
+            qmc_engine = self._cached_qmc_engine
+        else:
+            if self._qmc_type == "sobol":
+                qmc_engine = scipy.stats.qmc.Sobol(d, seed=self._seed, scramble=self._scramble)
+            elif self._qmc_type == "halton":
+                qmc_engine = scipy.stats.qmc.Halton(d, seed=self._seed, scramble=self._scramble)
+            else:
+                message = (
+                    f"The `qmc_type`, {self._qmc_type}, is not a valid. "
+                    'It must be one of "sobol" and "halton".'
+                )
+                raise ValueError(message)
+
+        assert isinstance(qmc_engine, scipy.stats.qmc.QMCEngine)
+
+        forward_size = sample_id - qmc_engine.num_generated  # `sample_id` starts from 0.
+        qmc_engine.fast_forward(forward_size)
```

Comment:
Bear in mind that this needs attention for Sobol'. If you want to fast forward what you can do is:

1. Fast forward to 2**m
2. Take 2**k sample with k <= m

Something else than this would break the properties. So if the function calling this respect these rules it's fine.

Reply:
In Optuna, it is beneficial to be able to one sample after another rather than sampler 2^k points at a time. Such a design simplifies the user interface and is compatible with existing APIs for sampling hyperparameters. For example, users often want to sample as many points as they can in a given time budget but are not sure how long evaluation of each hyperparameter takes. In this case, incremental sampling is beneficial because the users do not have to stop the sampling manually on their own.

I believe that we need to add to the document a caveat for the size of Sobol samples but I do not think it is a good idea to change the logic here.

Reply:
Exactly, just be clear in the notes about this. For this reason, I would default to `Halton`

Reply:
OK, I will make `Halton` the default. (I took benchmark and there was almost no difference in performance between Halton and Sobol.